src/structures/APIMessage.js

```diff
@@ -0,0 +1,321 @@
+
+    let mentionPart = '';
+    if (this.options.reply && !this.isUser && this.target.type !== 'dm') {
+      const id = this.target.client.users.resolveID(this.options.reply);
+      mentionPart = `<@${this.options.reply instanceof GuildMember && this.options.reply.nickname ? '!' : ''}${id}>, `;
+      if (isSplit) {
+        splitOptions.prepend = `${mentionPart}${splitOptions.prepend || ''}`;
+      }
+    }
+
+    if (content || mentionPart) {
+      if (isCode) {
+        const codeName = typeof this.options.code === 'string' ? this.options.code : '';
+        content = `${mentionPart}\`\`\`${codeName}\n${Util.escapeMarkdown(content, true)}\n\`\`\``;
+        if (isSplit) {
+          splitOptions.prepend = `${splitOptions.prepend || ''}\`\`\`${codeName}\n`;
+          splitOptions.append = `\n\`\`\`${splitOptions.append || ''}`;
+        }
+      } else if (mentionPart) {
+        content = `${mentionPart}${content}`;
+      }
+
+      const disableEveryone = typeof this.options.disableEveryone === 'undefined' ?
+        this.target.client.options.disableEveryone :
+        this.options.disableEveryone;
+      if (disableEveryone) {
+        content = content.replace(/@(everyone|here)/g, '@\u200b$1');
+      }
+
+      if (isSplit) {
+        content = Util.splitMessage(content, splitOptions);
+      }
+    }
+
+    return content;
+  }
+
+  /**
+   * Resolves both data and files.
+   * @returns {APIMessage}
+   */
+  async resolve() {
+    this.resolveData();
+    await this.resolveFiles();
+    return this;
+  }
+
+  /**
+   * Resolves data.
+   * @returns {APIMessage}
+   */
+  resolveData() {
+    if (this.data) {
+      return this;
+    }
+
+    const content = this.makeContent();
+    const tts = Boolean(this.options.tts);
+    let nonce;
+    if (typeof this.options.nonce !== 'undefined') {
+      nonce = parseInt(this.options.nonce);
+      if (isNaN(nonce) || nonce < 0) throw new RangeError('MESSAGE_NONCE_TYPE');
+    }
+
+    const embedLikes = [];
+    if (this.isWebhook) {
+      if (this.options.embeds) {
+        embedLikes.push(...this.options.embeds);
+      }
+    } else if (this.options.embed) {
+      embedLikes.push(this.options.embed);
+    }
+    const embeds = embedLikes.map(e => new MessageEmbed(e)._apiTransform());
+
+    let username;
+    let avatarURL;
+    if (this.isWebhook) {
+      username = this.options.username || this.target.name;
+      if (this.options.avatarURL) avatarURL = this.options.avatarURL;
+    }
+
+    const data = {
+      content,
+      tts,
+      nonce,
+      embed: this.options.embed === null ? null : embeds[0],
+      embeds,
+      username,
+      avatar_url: avatarURL,
+    };
+
+    this.data = data;
+    return this;
+  }
+
+  /**
+   * Resolves files.
+   * @returns {Promise<APIMessage>}
+   */
+  async resolveFiles() {
+    if (this.files) {
+      return this;
+    }
+
+    const embedLikes = [];
```

Comment:
Since `resolveData` is called before `resolveFiles`, you could remove a lot of duplication by reusing `this.data.embeds` and `this.data.embed`.